File: 703. Kth Largest Element in a Stream/main.js
// Still working with javascript, don't know it still be false
/**
 * @param {number} k
 * @param {number[]} nums
 */
var KthLargest = function (k, nums) {
  this.k = k;
  nums.sort();
  console.log("🚀 ~ KthLargest ~ nums:", nums);

  this.nums = nums;
};

KthLargest.prototype.indexBinarySearch = function (array, number) {
  let start = 0;
  let end = array.length - 1;

  while (start <= end) {
    let mid = Math.floor((start + end) / 2);

    // If K is found
    if (array[mid] == number) return mid;
    else if (array[mid] < number) start = mid + 1;
    else end = mid - 1;
  }

  // Return insert position
  return end + 1;
};

/**
 * @param {number} val
 * @return {number}
 */
KthLargest.prototype.add = function (val) {
  const index = this.indexBinarySearch(this.nums, val);
  console.log("🚀 ~ index:", index);
  this.nums.splice(index, 0, val);
  console.log("🚀 ~ this.nums:", this.nums);

  return this.nums[this.nums.length - this.k];
};

/**
 * Your KthLargest object will be instantiated and called as such:
 * var obj = new KthLargest(k, nums)
 * var param_1 = obj.add(val)
 */

const obj = new KthLargest(3, [4, 5, 8, 2]);
obj.add(3);
obj.add(5);
obj.add(10);
obj.add(9);
const param_2 = obj.add(4);
console.log("🚀 ~ param_1:", param_2);

File: 703. Kth Largest Element in a Stream/main.py
import heapq
from typing import List


class KthLargest:
    # Use binary search to find the position that number would be in the sorted array
    def binary_search_insert_position(self, array, number):
        left, right = 0, len(array) -1
        
        while left <= right:
            mid = (left + right) // 2
            if array[mid] < number:
                left = mid + 1
            elif array[mid] == number:
                return mid
            else:
                right = mid -1 
        
        return left 

    def __init__(self, k: int, nums: List[int]):
        self.k = k
        nums.sort()
        self.nums = nums
        

    def add(self, val: int) -> int:
        # Find the correct position using binary search
        position = self.binary_search_insert_position(self.nums, val)
        # Insert the number at the correct position
        self.nums.insert(position, val)
        return self.nums[-self.k]
    
    
    # Heap approach
    
    def __init__(self, k, nums):
        self.k = k
        self.nums = nums
        heapq.heapify(self.nums)  # Convert nums into a heap
        # If the heap is larger than k, remove the smallest elements until it has exactly k elements
        while len(self.nums) > k:
            heapq.heappop(self.nums)

    def add(self, val):
        heapq.heappush(self.nums, val)
        # If after adding the new value, the heap has more than k elements, pop the smallest
        if len(self.nums) > self.k:
            heapq.heappop(self.nums)
        # The smallest element in the heap is now the k-th largest element
        return self.nums[0]

        


# Your KthLargest object will be instantiated and called as such:
if __name__ == "__main__":
    obj = KthLargest(3, [4,5,8,2])
    param_1 = obj.add(3)
    print(param_1)